use-cases/rag-pipeline/alloy-db-setup/src/create_catalog.py

```diff
@@ -81,73 +69,100 @@ def create_database(database, new_database):
                     )
                     db_conn.execute(create_scann_extn)
-                    logger.info(
+                    logging.info(
                         f"alloydb_scann extension enabled successfully on db {new_database}."
                     )
     except Exception as e:
         logging.error(f"An error occurred while enabling the extensions: {e}")
     finally:
         if db_conn:
             db_conn.close()
-            logger.info(f"DB: {new_database} Connection closed")
+            logging.info(f"DB: {new_database} Connection closed")
         if connector:
             connector.close()
-            logger.info("Connector closed")
+            logging.info("Connector closed")
 
 
-def create_and_populate_table(database, table_name, processed_data_path):
-    """Creates and populates a table in PostgreSQL using pandas and sqlalchemy."""
-
+async def create_and_populate_table(
+    database, table_name, processed_data_path, max_workers_value
+):
+    """Creates and populates table, generating embeddings concurrently."""
     try:
-        # 1. Extract
+        # 1. Extract Data
         df = pd.read_csv(processed_data_path)
         logger.info(f"Input df shape: {df.shape}")
 
-        # Dropping products with image_uri as NaN
+        # Drop the products with image_uri as NaN
         df.dropna(subset=["image_uri"], inplace=True)
         logger.info(f"resulting df shape: {df.shape}")
 
-        logger.info(f"Starting embedding generation...")
-        # 2. Transform
-        logger.info(f"Starting multimodal embedding generation...")
-        df["multimodal_embeddings"] = df.apply(
-            lambda row: get_emb.get_embeddings(row["image_uri"], row["Description"]),
-            axis=1,
-        )
-        logger.info(f"Multimodal embedding generation completed")
+        # 2. Transform: Embedding Generation (aiohttp)
+        num_rows = len(df)
+        embedding_tasks = []
+        logger.info("Starting embedding generation...")
 
-        logger.info(f"Starting text embedding generation...")
-        df["text_embeddings"] = df.apply(
-            lambda row: get_emb.get_embeddings(None, row["Description"]), axis=1
+        # ClientSession outside loop for connection reuse. Timeout included.
+        timeout_settings = aiohttp.ClientTimeout(
+            total=300, sock_connect=10, sock_read=60
         )
-        logger.info(f"Text embedding generation completed")
+        async with aiohttp.ClientSession(
+            connector=aiohttp.TCPConnector(limit=max_workers_value),
+            raise_for_status=True,
+            timeout=timeout_settings,
+        ) as session:
+            for i in range(num_rows):
+                row = df.iloc[i]
+                # Start all tasks concurrently for max performance
+                embedding_tasks.append(
+                    get_emb.get_embeddings_async(
+                        session, row["image_uri"], row["Description"], timeout_settings
```

Comment:
Add a comma at the end for consistency and formatting.

Reply:
resolved